sympy/core/compatibility.py

```diff
@@ -396,15 +396,14 @@ def as_int(n, strict=True):
     ...
     ValueError: ... is not an integer
     """
-    from sympy.core.numbers import Integer
     try:
-        if strict and not isinstance(n, SYMPY_INTS + (Integer,)):
+        if strict and n.__index__() != n:
             raise TypeError
         result = int(n)
         if result != n:
             raise TypeError
         return result
-    except TypeError:
+    except AttributeError:
         raise ValueError('%s is not an integer' % (n,))
```

Comment:
We shouldn't catch `AttributeError`.

Use the `operator.index()` function instead which will raise `TypeError` instead. We don't need to use `int` if `index` returns a value. There should be no need to check the value returned by index either:
https://www.python.org/dev/peps/pep-0357/